internal/scheduler/scheduler.go

```diff
@@ -310,6 +348,39 @@ func (s *Scheduler) syncState(ctx context.Context) ([]*jobdb.Job, error) {
 	return jobsToUpdate, nil
 }
 
+func (s *Scheduler) createSchedulingInfoWithNodeAntiAffinityForAttemptedRuns(job *jobdb.Job) (*schedulerobjects.JobSchedulingInfo, error) {
+	newSchedulingInfo := proto.Clone(job.JobSchedulingInfo()).(*schedulerobjects.JobSchedulingInfo)
+	podSchedulingRequirement := PodRequirementFromJobSchedulingInfo(newSchedulingInfo)
+	if podSchedulingRequirement == nil {
+		return nil, fmt.Errorf("no pod scheduling requirement found for job %s", job.GetId())
+	}
+	newAffinity := podSchedulingRequirement.Affinity
+	if newAffinity == nil {
+		newAffinity = &v1.Affinity{}
+	}
+
+	for _, run := range job.AllRuns() {
+		if run.RunAttempted() {
+			affinity.AddNodeAntiAffinity(newAffinity, s.nodeIdLabel, run.Node())
+		}
+	}
+	podSchedulingRequirement.Affinity = newAffinity
+	return newSchedulingInfo, nil
+}
+
+func (s *Scheduler) addNodeAntiAffinitiesForAttemptedRunsIfSchedulable(job *jobdb.Job) (*jobdb.Job, bool, error) {
+	schedulingInfoWithNodeAntiAffinity, err := s.createSchedulingInfoWithNodeAntiAffinityForAttemptedRuns(job)
+	if err != nil {
+		return nil, false, err
+	}
+	podSchedulingRequirement := PodRequirementFromJobSchedulingInfo(schedulingInfoWithNodeAntiAffinity)
+	if podSchedulingRequirement == nil {
+		return nil, false, fmt.Errorf("no pod scheduling requirement found for job %s", job.GetId())
+	}
+	isSchedulable, _ := s.submitChecker.CheckPodRequirements(podSchedulingRequirement)
```

Comment:
This is a bit dodgy- I think this function is being called from `syncState` which means that all replicas of the scheduler are executing this regardless of whether they are leader.  Previously all the code called by this function was deterministic so we could guarantee that all replicas would end up in the same state.  This code breaks that assumption which means we could lose a job.  For example:
- Leader executes this code and determines job is ok to requeue
- Follower executes this code and determines job should fail (and marks it as such)
- Leader leases the job to the executor
- Follower Becomes Leader and fails the job
- Job run is now orphaned

TBF this whole bit of code is awkward because the point of `syncState` is to make the internal in-memory model look like the database.  The re-queuing of jobs doesn't fit nicely into that because we're not just taking the state from the db, but rather using scheduler-specific business logic to modify the state and at that point you run the risk that different scheduler instances may make different decisions. I've gone over a couple of different ways of sorting this and the best I can come up with is:
- Always requeue  in `syncstate`, but leave the job run as failed
- Move the requeing logic into `generateUpdateMessages` which can fail the job if it determines it shouldn't be requeued.

I think this works, because it means that only the leader will be checking whether something can be requeued.

Reply:
I agree with this and we can move it around

Is there a reason we don't send a requeue message in generateUpdateMessages - and remove all the business logic from syncstate.

Then only leader can decide if we requeue or fail the job and the others just get their state from the db.

Reply:
I think a `requeue` message is probably the best solution, it would also make life easier for e.g. lookout which could then mark the state correctly without having to gues what the scheduler has decided. 

The only slightly tricky bit is what to store in the schedulerDb as a bool won't be enough.  I think an int giving the number times it has been requeued should be enough, but I haven't thought through this enough to be sure.

Reply:
I do wonder if the requeue message could include the updated scheduling requirements.  That would be nice from the Pov of ensuing that all schedulers agree what nodes are being excluded.

Reply:
Are you envisioning that we'd conditionally exclude nodes?

Currently its either exclude all nodes we've tried on or fail - so it shouldn't be ambiguous  - however I'm happy to be proven wrong.

Although including the scheduling requirements on the requeue message would work and be pretty flexible

Reply:
So you're right- at the moment the logic is unambiguous so each scheduler can determine which nodes to exclude purely from the runs.  My thinking behind suggesting an explicit update to the scheduling requirements is mainly based around the feeling that the fewer decisions non-leader schedulers make, the less chance there is of them getting out of sync.  There's a few secondary benefits of this approach too- e.g. something like lookout could now show excluded nodes, but I'm not sure these are compelling enough to make a change.

The other thing I'm thinking of is how we integrate this with the hash-based scheduling optimization, which should go a long way toward mitigating the performance of `scheduleMany` as discussed in #2319.  Ideally we don't always compute hashes in `syncState` because this will involve computing `len(queue)` hashes every time the leader fails over.  Instead, I was hoping to calculate them in the ingester and a `requeued` message with an updated schedulingRequirements would be useful for this I think.

Reply:
I think that is all reasonable and won't be a lot more work.

In terms of implementation - would the plan be to just updated the scheduling_info column on the job table - on each `requeue` message that changes the scheduling  requirements?
 - Or would we want to record these updates elsewhere?
 - One pain point here is working out when the scheduling_info has changed in a cheap way - maybe this would be a reason to implement the hash in this PR

Would you be happy if I updated this PR to:
 - Add `requeue` event
 - Move all requeue logic to generateUpdateMessages (i.e send requeue event) and simplify `syncState`

Unsure if I want to also add a scheduling info hash in this PR - it depends the answer to the above

Reply:
> Add requeue event
> Move all requeue logic to generateUpdateMessages (i.e send requeue event) and simplify syncState

That seems good.  My suggestion for working out whether scheduling info has changed would be to add a seq no to scheduling info (both the table and the in-mem representation).  Every time you generate a new scheduling infor, increment the seq no and include it in the  `requeue` event.  The ingester should only overwrite the scheduling info if the seq num is higher (I think you can do this relatively cheaply with an ON CONFLICT), likewise the in memory representation should only be overwritten if the seq num is higher.

With regard for hashing- I think a tentative plan would be:
- Have a hash col that lives alongside the scheduling info and is versioned with it
- Scheduler ingester calculates and inserts the initial hash on the submit message
- Requeue message needs to include a hash (the leader scheduler will have to calculate the has when it requeues). The logic for updating the hash (both for the ingester and for follower schedulers) is identical to updating the scheduling info.

If you can get the hashing in that would be great, but if not it shouldn't be too hard to get it in later.

Reply:
> Add requeue event
> Move all requeue logic to generateUpdateMessages (i.e send requeue event) and simplify syncState

That seems good.  My suggestion for working out whether scheduling info has changed would be to add a seq no to scheduling info (both the table and the in-mem representation).  Every time you generate a new scheduling infor, increment the seq no and include it in the  `requeue` event.  The ingester should only overwrite the scheduling info if the seq num is higher (I think you can do this relatively cheaply with an ON CONFLICT), likewise the in memory representation should only be overwritten if the seq num is higher.

With regard for hashing- I think a tentative plan would be:
- Have a hash col that lives alongside the scheduling info and is versioned with it
- Scheduler ingester calculates and inserts the initial hash on the submit message
- Requeue message needs to include a hash (the leader scheduler will have to calculate the has when it requeues). The logic for updating the hash (both for the ingester and for follower schedulers) is identical to updating the scheduling info.

If you can get the hashing in that would be great, but if not it shouldn't be too hard to get it in later.